lib/Actions.js

```diff
@@ -147,76 +149,94 @@ Actions.prototype = {
                 type: this.TYPES.WAIT_FOR_QUERY
             });
-            waitingForQuery = this._performAction(waitAction, parentSelector, previousActionResult);
-        } else {
-            waitingForQuery = vow.resolve();
+            waitPromise = this.performAction(waitAction, parentSelector, prevResult);
         }
 
-        var casesPromise;
+        let casesPromise;
         if (action.cases) {
-            casesPromise = this._performCases(action.cases, parentSelector, previousActionResult);
+            debug('action.cases');
+            casesPromise = this._createInstance(
+                {
+                    type: 'cases',
+                    cases: action.cases
+                },
+                selector,
+                parentSelector,
+                prevResult
+            ).perform();
         }
 
         let actionPromise;
+
+        // mutation for if-then-else action
         if (action.conditions) {
-            actionPromise = this.performConditionalActions(selector, action.conditions, action.actions, action.elseActions);
-        } else {
-            const actionInstance = actionsFactory.createAction({
-                selector,
-                actionOptions: action,
-                env: this._env,
-                parser: this._parser,
-                actions: this,
-                previousActionResult
-            });
+            action.type = 'condition';
+        }
 
-            if (!actionInstance) {
-                return vow.reject(new Error('Unknown action type: ' + action.type));
-            }
+        const actionInstance = this._createInstance(action, selector, parentSelector, prevResult);
 
-            actionPromise = actionInstance.perform();
+        if (!actionInstance) {
+            return vow.reject(new Error('Unknown action type: ' + action.type));
         }
 
+        actionPromise = actionInstance.perform();
+
         if (action.transform) {
-            actionPromise = actionPromise.then(results => this._parser.transform(results, action.transform));
+            actionPromise = actionPromise.then(result => this._parser.transform(result, action.transform));
         }
 
         if (action.set) {
-            actionPromise = actionPromise.then(results => {
-                this.set(action.set, results);
-                return results;
+            actionPromise = actionPromise.then(result => {
+                this.set(action.set, result);
+                return result;
             });
         }
 
         return vow
-            .all([actionPromise, waitingForPage, waitingForQuery])
+            .all([actionPromise, waitPromise])
             .spread(result => casesPromise || result);
     },
 
     /**
+     * @param action
+     * @param selector
+     * @param parentSelector
+     * @param prevResult
+     * @return {Action}
+     * @private
+     */
+    _createInstance(action, selector, parentSelector, prevResult) {
```

Comment:
`_createInstance` is ambiguous name, maybe `_createAction` or `_createActionInstance` ?